### DISCUSSION TITLE
More cohesive `$for` introspection & unrolling by the parser

### DISCUSSION THREAD

**Discussion by ntrel** [vlang/v#11802] [Ideas]
First, suppose we could introspect fields directly without `$for`:
```v
struct S1 {
	i int
	s string
}

assert S1.i.name == 'i'
// note: typeof will be a type in future, not a string
// so typeof would naturally work with `is`
assert typeof(S1.i) is int
```
The above feature is not intended to be directly useful in source code but is necessary for implementing the introspection idea below.

Now, what if the `$for` 'loop' is unrolled by the *parser*. This happens before the checker and cgen see any statements. This means type checking and code generation just work naturally without special code for `$for`.

The parser treats `f` below as a symbol alias for each field. That alias is inlined when the body statements are unrolled:
```v
struct S1 {
	i int
	s string
}
// syntax: S1.fields could conflict with a field named `fields`, so use $fields
$for f in S1.$fields {
	// `f` is not a variable
	// `f` is an alias: first S1.i, second S1.s
	println(f.name)
	// typeof on an alias gives the type of the underlying aliased symbol
	println(typeof(f).name)
}
```
The parser generates an AST for 2 * 2 println statements:
```v
// 1st field
println(S1.i.name)
println(typeof(S1.i).name)
// 2nd field
println(S1.s.name)
println(typeof(S1.s).name)
```
Output:
```
i
int
s
string
```
Before we get to method iteration, let's look at introspecting a function and its parameters:
```v
fn f(i int, b bool)

assert f.name == 'f'
assert typeof(f) is fn(int, bool)
// f.params is a list of symbol aliases
// like an array, a symbol list has length and indexing
assert f.params.len == 2
// f.params[0] is an alias for f's `i` parameter
assert f.params[0].name == 'i'
// f.params[1] is an alias for f's `b` parameter
assert typeof(f.params[1]) is bool
```
Like a function, a method also has a list of symbol aliases to its parameters:
```v
fn (s S1) m1(i int, s string)
fn (s S1) m2(b bool)

assert S1.m1.name == 'm1'
// `params` includes the receiver
assert S1.m1.params[0].name == 's'
assert typeof(S1.m2.params[1]) is bool
```
Iterating methods is similar to iterating fields:
```v
$for m in S1.$methods {
	println(m.name)
	println(m.params.len)
}
```
The parser generates an AST for 2 * 2 println statements:
```v
println(S1.m1.name)
println(S1.m1.params.len)
println(S1.m2.name)
println(S1.m2.params.len)
```

**Comment by danieldaeschle** [comment]
I don't get what's new?

**Comment by ntrel** [comment]
> I don't get what's new?

> $for f in S1.$fields

That e.g. `f` is a symbolic alias, it is not a metadata struct instance like now. So we can use `typeof` on it to get the field type. Also that the parser handles `$for`, the checker never sees it directly, nevermind cgen.

**Comment by ntrel** [comment]
> S1.i.name == 'i'

I think to avoid conflicts with type properties e.g. (when there's a field called `name`) `S1.name == 'S1'`, we should use a different syntax like:
```v
S1.field('i').name // generated by the parser rather than written by a human
S1.method('m1').params.len
```
So `Type.fields` would continue working, no need for `Type.$fields`.

**Comment by danieldaeschle** [comment]
How will we handle autocompletion in vls? This the typesymbol accessable of this? with method `field()` etc?

**Comment by Ben-Fields** [comment]
> no need for `Type.$fields`

I actually *really like* the `$` prefix in the original proposal. I read this proposal before and searched a while to find it again.

I agree with:

> could conflict with a field named `fields`, so use $fields

`.$fields` clearly distinguishes it from regular struct fields and observes the established semantic convention of `$`. It also keeps things simple.

My name is Fields so clearly I have authority on this matter.

Overall a great proposal.
